src/librustc/infer/error_reporting/util.rs

```diff
@@ -0,0 +1,129 @@
+    // with 'a, the corresponding bound region and is_first which is true if
+    // the hir::Arg is the first argument in the function declaration.
+    pub fn find_arg_with_anonymous_region
+        (&self,
+         anon_region: Region<'tcx>,
+         replace_region: Region<'tcx>)
+         -> Option<(&hir::Arg, ty::Ty<'tcx>, ty::BoundRegion, bool)> {
+
+        match *anon_region {
+            ty::ReFree(ref free_region) => {
+
+                let id = free_region.scope;
+                let hir = &self.tcx.hir;
+                if let Some(node_id) = hir.as_local_node_id(id) {
+                    if let Some(body_id) = hir.maybe_body_owned_by(node_id) {
+                        let body = hir.body(body_id);
+                        if let Some(tables) = self.in_progress_tables {
+                            body.arguments
+                                .iter()
+                                .enumerate()
+                                .filter_map(|(index, arg)| {
+                                    let ty = tables.borrow().node_id_to_type(arg.id);
+                                    let mut found_anon_region = false;
+                                    let new_arg_ty = self.tcx
+                                        .fold_regions(&ty,
+                                                      &mut false,
+                                                      |r, _| if *r == *anon_region {
+                                                          found_anon_region = true;
+                                                          replace_region
+                                                      } else {
+                                                          r
+                                                      });
+                                    if found_anon_region {
+                                        let is_first = index == 0;
+                                        Some((arg, new_arg_ty, free_region.bound_region, is_first))
+                                    } else {
+                                        None
+                                    }
+                                })
+                                .next()
+                        } else {
+                            None
+                        }
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            }
+            _ => None,
+
+        }
+    }
+
+    // This method returns whether the given Region is Anonymous
+    // and returns the DefId and the BoundRegion corresponding to the given region.
+    pub fn is_suitable_anonymous_region(&self,
+                                        region: Region<'tcx>)
+                                        -> Option<(DefId, ty::BoundRegion)> {
+        match *region {
+            ty::ReFree(ref free_region) => {
```

Comment:
When `match`ing on only one arm it is a good idea to use `if let` instead, as it communicates intent to not deal with the other possible matches, as well as yielding fewer lines of code by not having to add `_ => ()`.